plugins/outputs/nats/nats.go

```diff
@@ -85,8 +128,124 @@ func (n *NATS) Connect() error {
 
 	// try and connect
 	n.conn, err = nats.Connect(strings.Join(n.Servers, ","), opts...)
+	if err != nil {
+		return err
+	}
 
-	return err
+	if n.Jetstream != nil {
+		n.jetstreamClient, err = jetstream.New(n.conn)
+		if err != nil {
+			return fmt.Errorf("failed to connect to jetstream: %w", err)
+		}
+
+		if len(n.Jetstream.Subjects) == 0 {
+			n.Jetstream.Subjects = []string{n.Subject}
+		}
+		// If the overall-subject is already present anywhere in the Jetstream subject we go from there,
+		// otherwise we should append the overall-subject as the last element.
+		if !choice.Contains(n.Subject, n.Jetstream.Subjects) {
+			n.Jetstream.Subjects = append(n.Jetstream.Subjects, n.Subject)
+		}
+		streamConfig, err := n.getJetstreamConfig()
+		if err != nil {
+			return fmt.Errorf("failed to parse jetstream config: %w", err)
+		}
+		_, err = n.jetstreamClient.CreateOrUpdateStream(context.Background(), *streamConfig)
+		if err != nil {
+			return fmt.Errorf("failed to create or update stream: %w", err)
+		}
+		n.Log.Infof("Stream (%s) successfully created or updated", n.Jetstream.Name)
+	}
+	return nil
+}
+
+func (n *NATS) getJetstreamConfig() (*jetstream.StreamConfig, error) {
+	var retention jetstream.RetentionPolicy
+	switch n.Jetstream.Retention {
+	case "", "limits":
+		retention = jetstream.LimitsPolicy
+	case "interest":
+		retention = jetstream.InterestPolicy
+	case "workqueue":
+		retention = jetstream.WorkQueuePolicy
+	default:
+		return nil, fmt.Errorf("invalid 'retention' setting %q", n.Jetstream.Retention)
+	}
+
+	var discard jetstream.DiscardPolicy
+	switch n.Jetstream.Discard {
+	case "", "old":
+		discard = jetstream.DiscardOld
+	case "new":
+		discard = jetstream.DiscardNew
+	default:
+		return nil, fmt.Errorf("invalid 'discard' setting %q", n.Jetstream.Discard)
+	}
+
+	var storage jetstream.StorageType
+	switch n.Jetstream.Storage {
+	case "memory":
+		storage = jetstream.MemoryStorage
+	case "", "file":
+		storage = jetstream.FileStorage
+	default:
+		return nil, fmt.Errorf("invalid 'storage' setting %q", n.Jetstream.Storage)
+	}
+
+	var compression jetstream.StoreCompression
+	switch n.Jetstream.Compression {
+	case "s2":
+		compression = jetstream.S2Compression
+	case "", "none":
+		compression = jetstream.NoCompression
+	default:
+		return nil, fmt.Errorf("invalid 'compression' setting %q", n.Jetstream.Compression)
+	}
```

Comment:
These should probably also be checked in `Init`.

Reply:
I agree.

Reply:
Done